### PR TITLE
Printing system state refactoring between Chip and Process

### PR DESCRIPTION
**Pull request by phil-levis**
### Pull Request Overview

This pull request refactors the boundaries for printing out system state. Currently, for ergonomics, printing out machine state requires a reference to a `Process`, which is a problem on systems that have no processes (all logic is in the kernel). This PR solves this problem as part of a general refactoring out printing out system state:

This PR breaks out printing state into four categories:
 - the `Chip` trait now has a `print_state` method that prints the control state of the MCU,
 - the `UserspaceKernelBoundary` trait has a method `print_context` which prints out the execution state of a stored context and stack,
 - the `Process` trait has two methods, `print_memory_map`, which prints the memory regions of a process, and `print_full_process`, which prints the full process state (execution state, memory map, and how the MPU is configured),
 - the MPU implementation implements `Display` which lets it print out the MPU state.

The one major implication of this refactoring is that the panic handler in a board's `io.rs` file needs access to a reference to `Chip`, so boards now store one in an Option called `CHIP`. 

This refactoring allows the kernel to print out machine state even when there are no processes.

### Testing Strategy

This pull request was tested by compiling and running on OpenTitan and compiling for all baords. OpenTitan has a forced load access fault in boot to show the debug output.

### TODO or Help Wanted

Testing. @bradjc has said that how the SCB registers on cortex-m are stored needs to be updated.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

**Comment by phil-levis-google** [comment]
> Great start, two things:
> 
> 1. cortex-m/lib.rs needs to be updated to store the SCB_REGISTERS state in both an app fault and a kernel fault.

Can you point me at the best way to do this?

> 2. The printing should move the arch crate, so multiple chips can call it without having to duplicate all of the print code.

Yup, agreed. Rather, see my most recent commit. It refactors output into 4 parts: machine state (defined by chip, currently just calls the arch version), execution context (by UserspaceKernelBoundary), process memory map/state (by Process) and MPU state (by MPU).

**Comment by bradjc** [comment]
> Can you point me at the best way to do this?

I think the changes would just need to be moving

https://github.com/tock/tock/blob/b49c7c9d422ae8efaa8b2fcf29639f529f33dda6/arch/cortex-m4/src/lib.rs#L425-L437

to the assembly block above the `if` statement in the `pub unsafe extern "C" fn hard_fault_handler()` function. That way they are saved on all hard faults.

File: kernel/src/platform/mod.rs [hunk 1]
```diff
@@ -100,6 +101,18 @@ pub trait Chip {
...
+    /// writer. This does not print out the execution context
+    /// (data registers), as this depends on how they are stored;
+    /// that is implemented by
+    /// `syscall::UserspaceKernelBoundary::print_context`.
+    /// This also does not print out a process memory state,
+    /// that is implemented by `process::Process::print_memory_map`.
+    /// The MPU state is printed by the MPU's implementation of
+    /// the Display trait.
+    /// Used by panic.
+    unsafe fn print_state(&self, writer: &mut dyn Write);
```
Comment (bradjc) [inline kernel/src/platform/mod.rs]: This doesn't seem like a function that should be marked `unsafe`. It should only read state, and display it. Are there memory safety concerns there?

File: kernel/src/process.rs [hunk 2]
```diff
@@ -219,8 +219,13 @@ pub trait ProcessType {
...
 
-    unsafe fn fault_fmt(&self, writer: &mut dyn Write);
-    unsafe fn process_detail_fmt(&self, writer: &mut dyn Write);
+    /// Print out the memory map (Grant region, heap, stack, program
+    /// memory, BSS, and data sections) of this process.
+    unsafe fn print_memory_map(&self, writer: &mut dyn Write);
+
+    /// Print out the full state of the process: its memory map, its
+    /// context, and the state of the memory protection unit (MPU).
+    unsafe fn print_full_process(&self, writer: &mut dyn Write);
```
Comment (bradjc) [inline kernel/src/process.rs]: It's great to have more of a clear intent behind these functions.

File: boards/opentitan/src/main.rs [hunk 3]
```diff
@@ -209,6 +212,10 @@ pub unsafe fn reset_handler() {
         FAULT_RESPONSE,
         &process_mgmt_cap,
     );
-
+    let bad_ptr = 0x81200000 as *mut u8;
+    let val = *bad_ptr;
+    if val == 0 {
+        debug!("Dereferenced bad pointer.");
+    }
```
Comment (bradjc) [inline boards/opentitan/src/main.rs]: Probably want to remove this?

File: kernel/src/platform/mod.rs (hunk 1 above)
Comment (phil-levis) [reply kernel/src/platform/mod.rs]: The original one was unsafe so I assumed it was for a good reason and made this unsafe too. It does access system registers, after all. You might not want to leak that state.

File: kernel/src/process.rs (hunk 2 above)
Comment (phil-levis) [reply kernel/src/process.rs]: `print_full_process` is simply a renaming of `process_detail_fmt`. `print_memory_map` factors out the part of this function that's actually implemented in `Process`. That way you can choose to print the memory map without also having to print all of the other state.

File: boards/opentitan/src/main.rs (hunk 3 above)
Comment (phil-levis) [reply boards/opentitan/src/main.rs]: Yes, before merging. But having this lets you test what a printout looks like. :)

**Comment by phil-levis** [comment]
> > Can you point me at the best way to do this?
> 
> I think the changes would just need to be moving
> 
> https://github.com/tock/tock/blob/b49c7c9d422ae8efaa8b2fcf29639f529f33dda6/arch/cortex-m4/src/lib.rs#L425-L437
> 
> to the assembly block above the `if` statement in the `pub unsafe extern "C" fn hard_fault_handler()` function. That way they are saved on all hard faults.

How's this?

https://github.com/tock/tock/pull/1590/commits/8b56873c71bcb181659dfc1d7b1897f557eaeb8a

File: kernel/src/platform/mod.rs (hunk 1 above)
Comment (bradjc) [reply kernel/src/platform/mod.rs]: I'm interested in @alevy's take, but accessing system registers, if we want to protect that, seems like something a capability would restrict. The functions were marked unsafe, but they were set that way long before we had this notion of restricting `unsafe` to just memory and type safety issues. This is the first time we have actually reviewed these functions and this interface.

**Comment by alistair23** [comment]
Are you planning on tidying up the commits before merging?

**Comment by phil-levis** [comment]
> This PR is a drastic improvement, but I still have a question about the `Chip` trait interface in light of #1088 and our discussions about how to mark functions as restricted.

This is a good point -- let's figure out how to refactor this and systematically nail down what Chip should be able to do.

File: boards/opentitan/src/main.rs (hunk 3 above)
Comment (alistair23) [reply boards/opentitan/src/main.rs]: Make sure this is dropped.

**Comment by bradjc** [comment]
Should I not have marked as "approved"? I thought the pointer dereference would be removed before merging.

**Comment by alistair23** [comment]
I did the same thing. It's an easy thing to forget about and it's not caught by the Travis test.

Maybe for future we only mark things as approved once it's all set?

**Comment by phil-levis** [comment]
This was my fault. I promised that I would and I forgot. Next time I say I’ll do something like that don’t believe me.

**Comment by phil-levis** [comment]
Well, I hope you all got to see the wonderful dump message!

**Comment by alistair23** [comment]
I've seen them a few times now :)

**Pull request merged**